File: game/game.go
package game

type Game struct {
	size           int
	startingPieces int
}

type Color byte
type Kind byte
type Piece byte

const (
	White Color = 1 << 7
	Black Color = 0 << 7

	colorMask byte = 1 << 7

	Flat     Kind = 1
	Standing Kind = 2
	Capstone Kind = 3

	typeMask byte = 1<<2 - 1
)

func makePiece(color Color, kind Kind) Piece {
	return Piece(byte(color) | byte(kind))
}

func pieceColor(p Piece) Color {
	return Color(byte(p) & colorMask)
}

func pieceKind(p Piece) Kind {
	return Kind(byte(p) & typeMask)
}

func isRoad(p Piece) bool {
	return pieceKind(p) == Flat || pieceKind(p) == Capstone
}

type Square []Piece

const hasCap = 1 << 16

type Position struct {
	game        *Game
	whiteStones int
	blackStones int
	move        int
	board       []Square
}

func (p *Position) At(x, y int) Square {
	return p.board[y*p.game.size+x]
}

func (p *Position) set(x, y int, s Square) {
	p.board[y*p.game.size+x] = s
}

func (p *Position) ToMove() Color {
	if p.move%2 == 0 {
		return White
	}
	return Black
}

func (p *Position) GameOver() (over bool, winner Color) {
	if p, ok := p.hasRoad(); ok {
		return true, p
	}

	if p.whiteStones&^hasCap != 0 && p.blackStones&^hasCap != 0 {
		return false, White
	}

	return true, p.flatsWinner()
}

func (p *Position) roadAt(x, y int) (Color, bool) {
	sq := p.At(x, y)
	if len(sq) == 0 {
		return White, false
	}
	return pieceColor(sq[0]), isRoad(sq[0])
}

func (p *Position) hasRoad() (Color, bool) {
	s := p.game.size
	white, black := false, false
	reachable := make([]Piece, s*s)
	for x := 0; x < s; x++ {
		if c, ok := p.roadAt(x, 0); ok {
			reachable[x] = makePiece(c, Flat)
		}
	}
	for y := 1; y < s; y++ {
		for x := 0; x < s; x++ {
			c, ok := p.roadAt(x, y)
			if !ok {
				continue
			}
			if reachable[x+(y-1)*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
		}
		for x := 0; x < s; x++ {
			c, ok := p.roadAt(x, y)
			if !ok {
				continue
			}
			if x > 0 && reachable[x-1+y*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
			if x < s-1 && reachable[x+1+y*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
		}
	}

	for x := 0; x < s; x++ {
		r := reachable[x+(s-1)*s]
		if r == makePiece(White, Flat) {
			white = true
		}
		if r == makePiece(Black, Flat) {
			black = true
		}
	}

	for i := range reachable {
		reachable[i] = Piece(0)
	}
	for y := 0; y < s; y++ {
		if c, ok := p.roadAt(0, y); ok {
			reachable[y*s] = makePiece(c, Flat)
		}
	}
	for x := 1; x < s; x++ {
		for y := 0; y < s; y++ {
			c, ok := p.roadAt(x, y)
			if !ok {
				continue
			}
			if reachable[x-1+y*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
		}
		for y := 0; y < s; y++ {
			c, ok := p.roadAt(x, y)
			if !ok {
				continue
			}
			if y > 0 && reachable[x+(y-1)*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
			if y < s-1 && reachable[x+(y+1)*s] == makePiece(c, Flat) {
				reachable[x+y*s] = makePiece(c, Flat)
			}
		}
	}
	for y := 0; y < s; y++ {
		r := reachable[y*s+s-1]
		if r == makePiece(White, Flat) {
			white = true
		}
		if r == makePiece(Black, Flat) {
			black = true
		}
	}

	switch {
	case white && black:
		if p.ToMove() == White {
			return Black, true
		}
		return White, true
	case white:
		return White, true
	case black:
		return Black, true
	default:
		return White, false
	}
}

func (p *Position) flatsWinner() Color {
	cw, cb := 0, 0
	for i := 0; i < p.game.size*p.game.size; i++ {
		stack := p.board[i]
		if len(stack) > 0 {
			if pieceKind(stack[0]) == Flat {
				if pieceColor(stack[0]) == White {
					cw++
				} else {
					cb++
				}
			}
		}
	}
	if cw > cb {
		return White
	}
	return Black
}

File: python/tak/game.py
import attr
import enum

from . import moves
from . import pieces

@attr.s(frozen=True, slots=True)
class Config(object):
  size   = attr.ib(validator = attr.validators.instance_of(int))
  pieces = attr.ib(validator = attr.validators.optional(attr.validators.instance_of(int)),
                   default = None)
  capstones = attr.ib(validator = attr.validators.optional(attr.validators.instance_of(int)),
                      default = None)

  @property
  def flat_count(self):
    if self.pieces is not None:
      return self.pieces
    return self.DEFAULT_PIECES[self.size]

  @property
  def capstone_count(self):
    if self.capstones is not None:
      return self.capstones
    return self.DEFAULT_CAPS[self.size]

  DEFAULT_PIECES = [0, 0, 0, 10, 15, 21, 30, 40, 50]
  DEFAULT_CAPS   = [0, 0, 0, 0, 0, 1, 1, 1, 2]

@attr.s(frozen=True, slots=True)
class StoneCounts(object):
  stones = attr.ib()
  caps   = attr.ib()

class WinReason(enum.Enum):
  ROAD = 1
  FLATS = 2

@attr.s(frozen=True, slots=True)
class Position(object):
  size   = attr.ib()
  stones = attr.ib()
  ply    = attr.ib()
  board  = attr.ib()

  @classmethod
  def from_config(cls, config):
    size   = config.size
    stones = StoneCounts(stones = config.flat_count,
                         caps   = config.capstone_count)

    return cls(
      size = size,
      ply = 0,
      stones = (stones, stones),
      board = [[] for _ in range(size*size)]
    )

  @classmethod
  def from_squares(cls, cfg, squares, ply):
    if len(squares) != cfg.size * cfg.size:
      raise ValueError("Wrong board size")

    counts = ([0,0], [0,0])

    for sq in squares:
      for p in sq:
        if p.kind == pieces.Kind.CAPSTONE:
          counts[p.color.value][1] += 1
        else:
          counts[p.color.value][0] += 1

    stones = (StoneCounts(cfg.flat_count - counts[0][0],
                          cfg.capstone_count - counts[0][1]),
              StoneCounts(cfg.flat_count - counts[1][0],
                          cfg.capstone_count - counts[1][1]))
    return cls(
      size = cfg.size,
      ply = ply,

      stones = stones,
      board = squares,
    )

  def to_move(self):
    if self.ply % 2 == 0:
      return pieces.Color.WHITE
    return pieces.Color.BLACK

  def in_bounds(self, x, y):
    return (x >= 0 and x < self.size and
            y >= 0 and y < self.size)

  def winner(self):
    color = self.has_road()
    if color is not None:
      return (color, WinReason.ROAD)
    if all(self.board) or any((s.stones+s.caps) == 0 for s in self.stones):
      return (self.flats_winner(), WinReason.FLATS)
    return (None, None)

  def is_road(self, x, y):
    sq = self[x,y]
    return len(sq) > 0 and sq[0].is_road()

  def _walk(self, seeds, color, horiz):
    seen = set()
    q = list(seeds)
    while q:
      j = q.pop()
      if j in seen:
        continue
      seen.add(j)
      x, y = j
      if not self.in_bounds(x,y):
        continue

      if not self.is_road(x, y) or self[x,y][0].color != color:
        continue

      if horiz and x == self.size - 1:
        return True
      if (not horiz) and y == self.size - 1:
        return True

      q.append((x+1, y))
      q.append((x-1, y))
      q.append((x, y+1))
      q.append((x, y-1))

    return False

  def has_road(self):
    left = [(0, i) for i in range(self.size)]
    top = [(i, 0) for i in range(self.size)]

    w = (self._walk(left, pieces.Color.WHITE, True) or
         self._walk(top, pieces.Color.WHITE, False))
    b = (self._walk(left, pieces.Color.BLACK, True) or
         self._walk(top, pieces.Color.BLACK, False))
    if w and b:
      return self.to_move().flip()
    if w:
      return pieces.Color.WHITE
    if b:
      return pieces.Color.BLACK
    return None

  def flat_counts(self):
    w, b = 0, 0
    for sq in self.board:
      if len(sq) == 0 or sq[0].kind != pieces.Kind.FLAT:
        continue
      if sq[0].color == pieces.Color.WHITE:
        w += 1
      else:
        b += 1
    return (w, b)

  def flats_winner(self):
    w,b = self.flat_counts()
    if w > b:
      return pieces.Color.WHITE
    if w < b:
      return pieces.Color.BLACK
    return self.to_move().flip()

  def __getitem__(self, pos):
    x,y = pos
    return self.board[y * self.size + x]

  def move(self, m):
    delta = {
      'ply': self.ply + 1,
    }

    if m.type.is_slide():
      self._move_slide(m, delta)
    else:
      self._move_place(m, delta)

    return attr.evolve(self, **delta)

  def _move_place(self, m, delta):
    if self.ply < 2 and m.type != moves.MoveType.PLACE_FLAT:
      raise IllegalMove("Illegal opening")
    if self[m.x,m.y]:
      raise IllegalMove("Place on an occupied square")
    color = self.to_move()
    if self.ply < 2:
      color = color.flip()

    slot = 'stones'
    kind = pieces.Kind.FLAT
    if m.type == moves.MoveType.PLACE_CAPSTONE:
      slot = 'caps'
      kind = pieces.Kind.CAPSTONE
    elif m.type == moves.MoveType.PLACE_STANDING:
      kind = pieces.Kind.STANDING

    cs = self.stones[color.value]
    if getattr(cs, slot) <= 0:
      raise IllegalMove("not enough stones")
    newstones = attr.evolve(cs, **{slot: getattr(cs, slot) - 1})

    if color == pieces.Color.WHITE:
      delta['stones'] = (newstones, self.stones[1])
    else:
      delta['stones'] = (self.stones[0], newstones)

    newboard = list(self.board)
    newboard[m.x + m.y*self.size] = [pieces.Piece(color=color, kind=kind)]
    delta['board'] = newboard

  def _move_slide(self, m, delta):
    if self.ply < 2:
      raise IllegalMove("Illegal opening")

    stack = self[m.x, m.y]
    ndrop = sum(m.slides)

    if ndrop > self.size or len(stack) < ndrop:
      raise IllegalMove("picking up too many pieces")

    if ndrop < 1:
      raise IllegalMove("must pick up at least one stone")

    if stack[0].color != self.to_move():
      raise IllegalMove("can't move opponent's stack")

    newboard = list(self.board)
    delta['board'] = newboard

    x, y = m.x, m.y
    dx, dy = m.type.direction()
    carry = stack[:ndrop]

    newboard[x + y *self.size] = stack[ndrop:]
    for drop in m.slides:
      x += dx
      y += dy
      if not self.in_bounds(x,y):
        raise IllegalMove("slide out of bounds")
      i = x + y * self.size
      orig = self.board[i]
      if len(orig) > 0 and orig[0].kind == pieces.Kind.CAPSTONE:
        raise IllegalMove("slide onto a capstone")
      if len(orig) > 0 and orig[0].kind == pieces.Kind.STANDING:
        if carry[0].kind != pieces.Kind.CAPSTONE or len(carry) != 1:
          raise IllegalMove("slide onto a standing stone")
        orig = [pieces.Piece(orig[0].color, pieces.Kind.FLAT)] + orig[1:]

      newboard[i] = carry[-drop:] + orig
      carry = carry[:-drop]


class IllegalMove(Exception):
  pass

__all__ = ['Config', 'StoneCounts', 'Position', 'IllegalMove', 'WinReason']
